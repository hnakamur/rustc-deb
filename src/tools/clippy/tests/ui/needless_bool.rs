#![warn(clippy::needless_bool)]

use std::cell::Cell;

macro_rules! bool_comparison_trigger {
    ($($i:ident: $def:expr, $stb:expr );+  $(;)*) => (

        #[derive(Clone)]
        pub struct Trigger {
            $($i: (Cell<bool>, bool, bool)),+
        }

        #[allow(dead_code)]
        impl Trigger {
            pub fn trigger(&self, key: &str) -> bool {
                $(
                    if let stringify!($i) = key {
                        return self.$i.1 && self.$i.2 == $def;
                    }
                 )+
                false
            }
        }
    )
}

#[allow(clippy::if_same_then_else)]
fn main() {
    let x = true;
    let y = false;
    if x {
        true
    } else {
        true
    };
    if x {
        false
    } else {
        false
    };
    if x {
        true
    } else {
        false
    };
    if x {
        false
    } else {
        true
    };
    if x && y {
        false
    } else {
        true
    };
    if x {
        x
    } else {
        false
    }; // would also be questionable, but we don't catch this yet
    bool_ret(x);
    bool_ret2(x);
    bool_ret3(x);
    bool_ret5(x, x);
    bool_ret4(x);
    bool_ret6(x, x);
    needless_bool(x);
    needless_bool2(x);
    needless_bool3(x);
}

#[allow(clippy::if_same_then_else, clippy::needless_return)]
fn bool_ret(x: bool) -> bool {
    if x {
        return true;
    } else {
        return true;
    };
}

#[allow(clippy::if_same_then_else, clippy::needless_return)]
fn bool_ret2(x: bool) -> bool {
    if x {
        return false;
    } else {
        return false;
    };
}

#[allow(clippy::needless_return)]
fn bool_ret3(x: bool) -> bool {
    if x {
        return true;
    } else {
        return false;
    };
}

#[allow(clippy::needless_return)]
fn bool_ret5(x: bool, y: bool) -> bool {
    if x && y {
        return true;
    } else {
        return false;
    };
}

#[allow(clippy::needless_return)]
fn bool_ret4(x: bool) -> bool {
    if x {
        return false;
    } else {
        return true;
    };
}

#[allow(clippy::needless_return)]
fn bool_ret6(x: bool, y: bool) -> bool {
    if x && y {
        return false;
    } else {
        return true;
    };
}

fn needless_bool(x: bool) {
    if x == true {};
}

fn needless_bool2(x: bool) {
    if x == false {};
}

fn needless_bool3(x: bool) {
    bool_comparison_trigger! {
        test_one:   false, false;
        test_three: false, false;
        test_two:   true, true;
    }

    if x == true {};
    if x == false {};
}
