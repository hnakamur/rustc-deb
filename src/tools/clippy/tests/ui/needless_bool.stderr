error: this if-then-else expression will always return true
  --> $DIR/needless_bool.rs:31:5
   |
LL | /     if x {
LL | |         true
LL | |     } else {
LL | |         true
LL | |     };
   | |_____^
   |
   = note: `-D clippy::needless-bool` implied by `-D warnings`

error: this if-then-else expression will always return false
  --> $DIR/needless_bool.rs:36:5
   |
LL | /     if x {
LL | |         false
LL | |     } else {
LL | |         false
LL | |     };
   | |_____^

error: this if-then-else expression returns a bool literal
  --> $DIR/needless_bool.rs:41:5
   |
LL | /     if x {
LL | |         true
LL | |     } else {
LL | |         false
LL | |     };
   | |_____^ help: you can reduce it to: `x`

error: this if-then-else expression returns a bool literal
  --> $DIR/needless_bool.rs:46:5
   |
LL | /     if x {
LL | |         false
LL | |     } else {
LL | |         true
LL | |     };
   | |_____^ help: you can reduce it to: `!x`

error: this if-then-else expression returns a bool literal
  --> $DIR/needless_bool.rs:51:5
   |
LL | /     if x && y {
LL | |         false
LL | |     } else {
LL | |         true
LL | |     };
   | |_____^ help: you can reduce it to: `!(x && y)`

error: this if-then-else expression will always return true
  --> $DIR/needless_bool.rs:74:5
   |
LL | /     if x {
LL | |         return true;
LL | |     } else {
LL | |         return true;
LL | |     };
   | |_____^

error: this if-then-else expression will always return false
  --> $DIR/needless_bool.rs:83:5
   |
LL | /     if x {
LL | |         return false;
LL | |     } else {
LL | |         return false;
LL | |     };
   | |_____^

error: this if-then-else expression returns a bool literal
  --> $DIR/needless_bool.rs:92:5
   |
LL | /     if x {
LL | |         return true;
LL | |     } else {
LL | |         return false;
LL | |     };
   | |_____^ help: you can reduce it to: `return x`

error: this if-then-else expression returns a bool literal
  --> $DIR/needless_bool.rs:101:5
   |
LL | /     if x && y {
LL | |         return true;
LL | |     } else {
LL | |         return false;
LL | |     };
   | |_____^ help: you can reduce it to: `return x && y`

error: this if-then-else expression returns a bool literal
  --> $DIR/needless_bool.rs:110:5
   |
LL | /     if x {
LL | |         return false;
LL | |     } else {
LL | |         return true;
LL | |     };
   | |_____^ help: you can reduce it to: `return !x`

error: this if-then-else expression returns a bool literal
  --> $DIR/needless_bool.rs:119:5
   |
LL | /     if x && y {
LL | |         return false;
LL | |     } else {
LL | |         return true;
LL | |     };
   | |_____^ help: you can reduce it to: `return !(x && y)`

error: equality checks against true are unnecessary
  --> $DIR/needless_bool.rs:127:8
   |
LL |     if x == true {};
   |        ^^^^^^^^^ help: try simplifying it as shown: `x`
   |
   = note: `-D clippy::bool-comparison` implied by `-D warnings`

error: equality checks against false can be replaced by a negation
  --> $DIR/needless_bool.rs:131:8
   |
LL |     if x == false {};
   |        ^^^^^^^^^^ help: try simplifying it as shown: `!x`

error: equality checks against true are unnecessary
  --> $DIR/needless_bool.rs:141:8
   |
LL |     if x == true {};
   |        ^^^^^^^^^ help: try simplifying it as shown: `x`

error: equality checks against false can be replaced by a negation
  --> $DIR/needless_bool.rs:142:8
   |
LL |     if x == false {};
   |        ^^^^^^^^^^ help: try simplifying it as shown: `!x`

error: aborting due to 15 previous errors

